### PR TITLE
Memory: Use a table based lookup scheme to read from memory regions

### PR DESCRIPTION
**Pull request by yuriks**
This replaces the tree of if-else in the memmap functions with a table of pointers which can be updated during runtime.

This is mostly ground work for proper kernel address space management, but in the meantime, gives a small ~3% speed up in general.

### CODE REVIEW

File: src/core/hle/shared_page.h [hunk 1]
```diff
@@ -11,18 +11,45 @@
...
 
 ////////////////////////////////////////////////////////////////////////////////////////////////////
 
 namespace SharedPage {
 
-template <typename T>
-void Read(T &var, const u32 addr);
-
-void Set3DSlider(float amount);
+// see http://3dbrew.org/wiki/Configuration_Memory#Shared_Memory_Page_For_ARM11_Processes
```
Comment (bunnei) [inline src/core/hle/shared_page.h]: *See

File: src/core/hle/shared_page.h [hunk 2]
```diff
@@ -11,18 +11,45 @@
...
+
+struct DateTime {
+    u64_le date_time;                  // 0
+    u64_le update_tick;                // 8
+    INSERT_PADDING_BYTES(0x20 - 0x10); // 10
+};
+static_assert(sizeof(DateTime) == 0x20, "Datetime size is wrong");
+
+struct SharedPageDef {
+    // most of these names are taken from the 3dbrew page linked above.
```
Comment (bunnei) [inline src/core/hle/shared_page.h]: *Most (sorry, comments that aren't capitalized kill me, I know, this isn't your doing)

File: src/core/hle/shared_page.h [hunk 3]
```diff
@@ -11,18 +11,45 @@
...
+    u32_le   date_time_selector;         // 0
+    u8       running_hw;                 // 4
+    u8       mcu_hw_info;                // 5: don't know what the acronyms mean
```
Comment (bunnei) [inline src/core/hle/shared_page.h]: This is a terrible line... "unknown" and get rid of comment?

File: src/core/hle/shared_page.h [hunk 4]
```diff
@@ -11,18 +11,45 @@
...
+    INSERT_PADDING_BYTES(0x20 - 0x6);    // 6
+    DateTime date_time_0;                // 20
+    DateTime date_time_1;                // 40
+    u8       wifi_macaddr[6];            // 60
+    u8       wifi_unknown1;              // 66: 3dbrew says these are "Likely wifi hardware related"
```
Comment (bunnei) [inline src/core/hle/shared_page.h]: 3DBrew

File: src/core/hle/shared_page.h (hunk 3 above)
Comment (yuriks) [inline src/core/hle/shared_page.h]: I'm just gonna get rid of the comment, it's pretty obvious to me it means "microcontroller hardware info".

Comment (Subv) [inline src/core/hle/shared_page.h]: Please add a comment saying that, the MCU = Microcontroller Hardware Info wasn't obvious at all to me :P

File: src/core/memory.cpp [hunk 5]
```diff
@@ -0,0 +1,187 @@
+// Copyright 2014 Citra Emulator Project
```
Comment (Subv) [inline src/core/memory.cpp]: 2015

File: src/core/memory.h [hunk 6]
```diff
@@ -0,0 +1,125 @@
+// Copyright 2014 Citra Emulator Project
```
Comment (Subv) [inline src/core/memory.h]: 2015, same in most other files

File: src/core/memory.h [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+// Licensed under GPLv2 or any later version
+// Refer to the license.txt file included.
+
+#pragma once
+
+#include "common/common_types.h"
+
+namespace Memory {
+
+const u32 PAGE_SIZE = 0x1000;
```
Comment (Subv) [inline src/core/memory.h]: Is this arbitrary or is it an official value?

**Comment by Subv** [comment]
Overall LGTM after the commets are addressed

File: src/core/memory.h (hunk 7 above)
Comment (yuriks) [inline src/core/memory.h]: It's the page size used by the ARM architecture... (of the sub-pages, anyway, real ARM uses a 2-level pagetable.)

Comment (archshift) [inline src/core/memory.h]: It's an official value.

**Comment by yuriks** [comment]
Updated.

@Subv: I didn't change all the copyrights since most of those files we're just moved from existing code, and I don't think you should update the date on those situations, so I just changed it on `memory.cpp` which I mostly re-wrote from scratch. Maybe @neobrain has an opinion on it.

File: src/core/memory.cpp [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+const size_t KB = 1024;
+const size_t MB = 1024*KB;
+
+const u32 PAGE_MASK = PAGE_SIZE - 1;
+const int PAGE_BITS = 12;
+
+enum class PageType {
+    Unmapped = 0,
+    Memory = 1,
+    Special = 2,
```
Comment (bunnei) [inline src/core/memory.cpp]: What does "Special" mean, except that it's used below for IO?

**Comment by bunnei** [comment]
Looks good!

**Comment by bunnei** [comment]
This is pending some documentation before being ready for merge

**Comment by yuriks** [comment]
Updated. Someone just quickly proofread the last commit before I squash.

**Comment by Subv** [comment]
LGTM

**Pull request merged**